### ISSUE TITLE
 google_checks.xml does not allow valid lower camel case identifiers

### ISSUE THREAD

**Issue by jakub-bochenski** [checkstyle/checkstyle#6203]
There is an identifier regex of form `^[a-z]([a-z0-9][a-zA-Z0-9]*)?$` in several places in https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L149

This will not allow valid lower camel case form of expressions like "X Factor" (which would be "xFactor")

**Comment by rnveach** [comment]
This looks like a duplicate of https://github.com/checkstyle/checkstyle/issues/5504

**Comment by jakub-bochenski** [comment]
OK, so quoting @romani 

> I do not share with you your naming convention and I do not see cases where single character is reasonable and clear to read (I have not see whole world sources, so I limited to my experience only).

My actual use case is `X-Transaction` which is a common thing to handle in webapps (as well as other `X-` headers).

**Comment by romani** [comment]
@jakub-bochenski , looks like more and more people disagree with my vision of naming. It is ok.

Here we have google style, so it is not mine style, I will be ok to update only if we get clarification from Google on this.

https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names
> In Google Style, special prefixes or suffixes are not used. For example, these names are not Google Style: name_, mName, s_name and kName.

I agree that your case it kind of edge case.
Please do conversation with Google first on this (you can create issue https://github.com/google/styleguide/issues on them).
Ones they agree to change their style with extra nuance/allowance. I will be happy to change config on our side.

It is not first case when different people read style in different way, the only way out is clarification of style guide.

**Comment by jakub-bochenski** [comment]
Seeing that the previous [issue](https://github.com/google/styleguide/issues/326) hasn't received any responses since February I think it would be rather pointless to ask again.
Actually I don't think any clarification is needed here. 

The `x` in `xTransaction` is not a prefix, special or otherwise. It's just a result of applying camel case to a phrase starting with `X`.
The regexp is supposed to detect prefixes, but it's giving a false positive in this case. It's a bug in how the check is configured.

**Comment by rnveach** [comment]
> make suppression for all violations for this Check that starts from "x"

Just to point out, users who use google_checks.xml directly can't make suppressions because there are no filters in the config. They would have to copy/paste their own config to add it, which also means updating it when it changes in a release.

**Comment by romani** [comment]
I do believe that is possible to be done in some plugins where filters could be configured separately

**Comment by yarivy** [comment]
While I see no way of distinguishing prefixes from words, 'x' is clearly a word in some settings, such as "x-axis", and it appears as such in the [dictionary](https://en.oxforddictionaries.com/definition/x). Google's specification clearly specifies that "x-axis" should be converted to "xAxis" (see "check non-empty" example in note no camel case).

This test is overzealous with respect to the style guide, and it leaves no good solutions for some names. In our project we change the config to specifically allow 'x' and 'y' as single letter words (and would consider more if need arises).

**Comment by romani** [comment]
>Google's specification clearly specifies that "x-axis" should be converted to "xAxis" (see "check non-empty" example in note no camel case).

Please give strict link to let me recheck.

**Comment by yarivy** [comment]
"x-axis" is not given as an example, but I will point out the rule and the way it applies here. The reference is to [the definition of camel case](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case). If you accept "x-axis" as an acceptable "prose form of a name" and follow the scheme you get the following:
prose form: "x-axis"
after step 1: "x-axis"
after step 2: "x axis"
after step 3: "x Axis"
after step 4: "xAxis"

I would further say that the current regex implies that the "prose form of a name" can't have a single letter leading word, but still allows a single letter word in other places, which doesn't make sense to me.
And, again, I don't have a reasonable general solution for the issue. For our project we simply changed the regex to allow phrases beginning in 'x' or 'y' specifically, but this is not a good general solution.

**Comment by romani** [comment]
We will not come to agreement here, we need to bring that all to Google style issue tracker. Ones they change documentation, we will change a config.

For now it is better make a copy and adjust it to allow such names. There are other cases where style guide us not clear, and users read them in different ways.

**Comment by jakub-bochenski** [comment]
> Do you know how to distinguish where single character before capital is preffix and when is the is Camel case ?

@romani  this is like asking "how to parse HTML using regex". 
I don't get what should Google update in their documentation to change the above. 
The tool (regex) is not powerful enough for the task and will result in false positives.

You might still want to use a heuristic like this if it catches more problems than it causes, but this is a matter of opinion. 

I think the best solution would be to have to checstyle configs for Google Style Guide:
 - opinionated with heuristics like this
 -  minimal with as few false positives as possible (it could also incorporate workarounds for known issues like https://issues.apache.org/jira/browse/MCHECKSTYLE-354 or https://github.com/checkstyle/checkstyle/issues/3342)

**Comment by yarivy** [comment]
For future readers, this are the regexes I use. Adding more allowed characters (z is an obvious case) should be simple.
for MemberName:   ^[a-z][a-z0-9][a-zA-Z0-9]*$|^[xy]([a-zA-Z0-9]*)?$
for various *ParameterNames, and for LocalVariableName:   ^[a-z]([a-z0-9][a-zA-Z0-9]*)?$|^[xy][a-zA-Z0-9]*$

**Comment by romani** [comment]
> The tool (regex) is not powerful enough for the task and will result in false positives.

regexp is just existing tool already, it was ok at the beginning. 
We can make new Check with custom parser, and use more smart Check in config.

> opinionated with heuristics like this

It is already opinionated up to my style understanding of 2014 year :). But there was a storm of corrections and some of them are conflicting, so I changed position to follow what is written in style. Some corrections from us already applied to google style, some still need support from community. So please go to google tracker and share your vote.

> minimal with as few false positives as possible 

please do not hesitate to send PR :), if issue is approved and nobody assigned on it - it is waiting for somebody to fix it.
